noto-fonts-emoji: Package both color and monochrome variants

nicoonoclaste opened the pull request:
## Description of changes

- noto-fonts-emoji → noto-fonts-color-emoji
- noto-fonts-monochrome-emoji: init at v46.2023-09-07

Closes #221181.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

nicoonoclaste committed:
noto-fonts-emoji → noto-fonts-color-emoji

Clarify that the monochrome font is not included, per #221181.

The new name is also coherent with the name of the font,
according to `fontconfig`: Noto Color Emoji.

nicoonoclaste committed:
noto-fonts-monochrome-emoji: init at version 46.2023-09-07

yorickvP commented on pkgs/data/fonts/noto-fonts/default.nix:
```diff
@@ -217,14 +217,64 @@ rec {
+      preferLocalBuild = true;
+
+      dontUnpack = true;
+      srcs = with lib; let
```
- `with lib;` -> `lib.mapAttrsToList` below
- could you lift this out of the derivation to line 231?
- I think the manual alignment will have to go, even though it looks nice
- It may be better transposed, but feel free to ignore:
```nix
{
   "300" = {
        weight = "Light";
        hash = "sha256-AXGLdWebddyJhTKMW/D/6tW8ODcaXrUM96m2hN9wYlg=";
    };
}
````

yorickvP commented on pkgs/data/fonts/noto-fonts/default.nix:
```diff
@@ -217,14 +217,64 @@ rec {
+          "700"   = "Bold";
+        };
+        fileHashes = {
+          "NotoEmoji-Bold.ttf"     = "ce426e27c6254eb515fb6f301c8aa7cb7c90be3bd9a843c6e165d899a2dc63c0";
```
Please convert the hashes to SRI using `nix hash to-sri --type sha256`:

yorickvP commented on pkgs/data/fonts/noto-fonts/default.nix:
```diff
@@ -217,14 +217,64 @@ rec {
+    let metadata = with builtins; head (fromJSON (readFile ./noto-emoji.json)).items;
+
+    in
+    stdenvNoCC.mkDerivation rec {
```
```suggestion
    stdenvNoCC.mkDerivation {
```

yorickvP commented on pkgs/data/fonts/noto-fonts/default.nix:
```diff
@@ -217,14 +217,64 @@ rec {
+      meta = with lib; {
+        description = "Monochrome emoji font";
+        homepage = "https://fonts.google.com/noto/specimen/Noto+Emoji";
+        license = with licenses; [ ofl ];
```
```suggestion
        license = licenses.ofl;
```

yorickvP commented on pkgs/data/fonts/noto-fonts/default.nix:
```diff
@@ -217,14 +217,64 @@ rec {
+        };
+      in mapAttrsToList
+        (variant: url: fetchurl rec { name = "NotoEmoji-${weightNames.${variant}}.ttf";
+                                      sha256 = fileHashes.${name};
```
when converted to SRI:
```suggestion
                                      hash = fileHashes.${name};
```

yorickvP commented on pkgs/top-level/aliases.nix:
```diff
@@ -1224,6 +1224,7 @@ mapAliases ({
   nomad_1_3 = throw "nomad_1_3 has been removed because it's outdated. Use a a newer version instead"; # Added 2023-09-02
   nordic-polar = throw "nordic-polar was removed on 2021-05-27, now integrated in nordic"; # Added 2021-05-27
   noto-fonts-cjk = noto-fonts-cjk-sans; # Added 2021-12-16
+  noto-fonts-emoji = throw "'noto-fonts-emoji' was renamed to 'noto-fonts-color-emoji'";
```
why not make an alias (possibly with a warning)?

yorickvP reviewed (changes requested):
Thanks for this! I added some feedback.

nicoonoclaste committed:
noto-fonts-monochrome-emoji: Automatically collect file hashes

nicoonoclaste committed:
noto-fonts-monochrome-emoji: Automatically update metadata

nicoonoclaste committed:
noto-fonts-emoji-monochrome: Validate cached SRI hashes

This hedges against a buggy version of the script generating data
that is then blindly reused even after the bug is fixed.

Mathnerd314 commented:
>  Is there a good reason we can't drop a Google Fonts API key in the repo?

We do have a [Chromium key](https://github.com/NixOS/nixpkgs/blob/27eedb56010219cd7f779e87b9a00c7504f76125/pkgs/applications/networking/browsers/chromium/common.nix#L357). It looks like @aszlig got permission for it [many years ago](https://github.com/NixOS/nixpkgs/commit/3c60e2ec39b11353779518e39a1d4aad13193b99). As mentioned in that commit, the public Google API TOS does forbid distributing API keys publicly (https://developers.google.com/terms: "Developer credentials may not be embedded in open source projects."), so a Fonts API key would similarly require special permission from Google. I'm not sure who you would reach out to now.

aszlig commented:
> We do have a [Chromium key](https://github.com/NixOS/nixpkgs/blob/27eedb56010219cd7f779e87b9a00c7504f76125/pkgs/applications/networking/browsers/chromium/common.nix#L357). It looks like @aszlig got permission for it [many years ago](https://github.com/NixOS/nixpkgs/commit/3c60e2ec39b11353779518e39a1d4aad13193b99). As mentioned in that commit, the public Google API TOS does forbid distributing API keys publicly (https://developers.google.com/terms: "Developer credentials may not be embedded in open source projects."), so a Fonts API key would similarly require special permission from Google. I'm not sure who you would reach out to now.

Good luck... we're currently running into quota limits for these mentioned API keys (they're also used in [Firefox](https://github.com/NixOS/nixpkgs/blob/f49b186df24d55d8d16040a1c8b90a14ede76eb2/pkgs/applications/networking/browsers/firefox/common.nix#L354-L356) btw.) and so far we were unable to reach anyone from Google who's in charge of these after multiple attempts. If you do for some reason find the right contact at Google, please let us know (cc: @mweinelt). The person mentioned in that referenced commit no longer works there.

nicoonoclaste committed:
noto-fonts-monochrome-emoji: Handle 403 errors in update script

mweinelt commented:
Fixed in https://github.com/NixOS/nixpkgs/pull/254387. Please rebase.
